### PR TITLE
MDList

### PR DESCRIPTION
**Pull request by Beider**
MDList has been added along with the CommandReplicator which can be used for other similar implementations that relies on sending commands across the network.

This also contains the new `this.ChangeNetworkMaster(int PeerId)` command.

This does not yet close the issue for List and Dictionary replication as dictionary is not yet included. However this feature is already massive enough so it might be good to review this part now before I add the second part.

### CODE REVIEW

File: MDNetworking/MDGameSession.cs [hunk 1]
```diff
@@ -765,5 +765,34 @@ public void NotifyPlayerNameChanged(int peerId)
...
+
+        public bool ChangeNetworkMaster(Node Node, int NewNetworkMaster)
+        {
+            if (!GetAllPeerIds().Contains(NewNetworkMaster))
+            {
+                // Invalid network master
+                return false;
+            }
+            
+            // Only server can change network master, we could also allow server?
```
Comment (DoubleDeez) [inline MDNetworking/MDGameSession.cs]: confused by the second part of this comment. Is this meant to say we could also allow clients? I think keeping it server-only is a good idea

File: MDNetworking/MemberImpl/MDReplicatedCommandReplicator.cs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        }
+
+        public override void Replicate(int JoinInProgressPeerId, bool IsIntervalReplicationTime)
+        {
+            IMDCommandReplicator CommandReplicator = GetCommandReplicator();
+            Node Instance = NodeRef.GetRef() as Node;
+
+            if (CommandReplicator == null)
+            {
+                MDLog.Error("Command replicator is null for member {0}#{1}", Instance.GetPath(), Member.Name);
```
Comment (DoubleDeez) [inline MDNetworking/MemberImpl/MDReplicatedCommandReplicator.cs]: Missing LOG_CAT param here and you should use [string interpolation](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/interpolated)

File: MDNetworking/Wrappers/MDList.cs [hunk 3]
```diff
@@ -0,0 +1,745 @@
...
+
+            if (CmdNumber > CommandCounter)
+            {
+                CommandQueue.Add(new ListCommandRecord(CmdNumber, Type, Parameters));
+                return;
+            }
+            else if (CmdNumber < CommandCounter)
+            {
+                // This should not happen
+                MDLog.Error(LOG_CAT, "Recieved a command with number {0} when our internal CommandCounter is {1}", CmdNumber, CommandCounter);
```
Comment (DoubleDeez) [inline MDNetworking/Wrappers/MDList.cs]: Should use String Interpolation

File: MDNetworking/Wrappers/MDList.cs [hunk 4]
```diff
@@ -0,0 +1,745 @@
...
+        {
+            // Parse input
+            int CmdNumber = Convert.ToInt32(Params[0]);
+            MDListActions Type = (MDListActions)Enum.Parse(typeof(MDListActions), Params[1].ToString());
+            object[] Parameters = Params.SubArray(2);
+
+            MDLog.Trace(LOG_CAT, "Recieved command [{0}] {1}", CmdNumber.ToString(), Type.ToString());
+            foreach(object obj in Parameters)
+            {
+                MDLog.Trace(LOG_CAT, "Parameter: {0}", obj.ToString());
```
Comment (DoubleDeez) [inline MDNetworking/Wrappers/MDList.cs]: Should use String Interpolation

File: MDNetworking/Wrappers/MDList.cs [hunk 5]
```diff
@@ -0,0 +1,745 @@
...
+
+        ///<summary>Do not call this! This should only be used by the MDReplicator</summary>
+        public void MDProcessCommand(params object[] Params)
+        {
+            // Parse input
+            int CmdNumber = Convert.ToInt32(Params[0]);
+            MDListActions Type = (MDListActions)Enum.Parse(typeof(MDListActions), Params[1].ToString());
+            object[] Parameters = Params.SubArray(2);
+
+            MDLog.Trace(LOG_CAT, "Recieved command [{0}] {1}", CmdNumber.ToString(), Type.ToString());
```
Comment (DoubleDeez) [inline MDNetworking/Wrappers/MDList.cs]: Should use String Interpolation

File: MDNetworking/Wrappers/MDList.cs [hunk 6]
```diff
@@ -0,0 +1,745 @@
...
+        {
+            MDReplicatedSetting[] ComparatorList = MDReplicator.ParseParameters(typeof(MDList.Settings), Settings);
+            foreach (MDReplicatedSetting setting in ComparatorList)
+            {
+                switch ((MDList.Settings)setting.Key)
+                {
+                    case MDList.Settings.COMPARATOR:
+                        if (!(setting.Value is Type))
+                        {
+                            MDLog.Warn(LOG_CAT, "{0} is not a type, use typeof(class) as value when using MDReplicatedSetting of type MDComparators", setting.Value.ToString());
```
Comment (DoubleDeez) [inline MDNetworking/Wrappers/MDList.cs]: Should use String Interpolation

File: MDNetworking/Wrappers/MDList.cs [hunk 7]
```diff
@@ -0,0 +1,745 @@
...
+                        }
+                        Type settingType = (Type)setting.Value;
+                        IComparer<T> comparer = Activator.CreateInstance(settingType) as IComparer<T>;
+                        if (comparer != null)
+                        {
+                            AddComparer(comparer);
+                        }
+                        else
+                        {
+                            MDLog.Error(LOG_CAT, "{0} is not an IComparer<T>.", settingType.ToString());
```
Comment (DoubleDeez) [inline MDNetworking/Wrappers/MDList.cs]: Should use String Interpolation

File: MDNetworking/Wrappers/MDList.cs [hunk 8]
```diff
@@ -0,0 +1,745 @@
...
+                }
+            }
+        }
+
+        protected void AddComparer(IComparer<T> Comparator)
+        {
+            IComparer<T> FoundComparator = GetComparatorByType(Comparator.GetType());
+            if (FoundComparator != null)
+            {
+                MDLog.Warn(LOG_CAT, "Comparator with key {0} has already been registered", Comparator.GetType().ToString());
```
Comment (DoubleDeez) [inline MDNetworking/Wrappers/MDList.cs]: Should use String Interpolation

**Review by DoubleDeez — CHANGES_REQUESTED** [review]
Looks good, I think there were just issues with the Logs, one major issue and the rest just to change to string interpolation

File: MDNetworking/MDGameSession.cs (hunk 1 above)
Comment (Beider) [reply MDNetworking/MDGameSession.cs]: Yeah that was supposed to say client, but I don't think we want client to do that so I will remove that part of the comment.

**Comment by Beider** [comment]
I fixed the issues from the review, also found a string interpolation string in game session that was missing the $ so added that as well.

**Review by DoubleDeez — APPROVED** [review]

**Pull request merged**